Mathlib/Tactic/FinCases.lean

```diff
@@ -42,20 +42,27 @@ and we return a list of the first goals which appeared.
 This is useful for hypotheses of the form `h : a ∈ [l₁, l₂, ...]`,
 which will be transformed into a sequence of goals with hypotheses `h : a = l₁`, `h : a = l₂`,
 and so on.
+Cases are named according to the order in which they are generated as tracked by `counter`
+and prefixed with `userNamePre`.
 -/
-partial def unfoldCases (g : MVarId) (h : FVarId) : MetaM (List MVarId) := do
+partial def unfoldCases (g : MVarId) (h : FVarId)
+    (userNamePre : Name := .anonymous) (counter := 0) : MetaM (List MVarId) := do
   let gs ← g.cases h
   try
     let #[g₁, g₂] := gs | throwError "unexpected number of cases"
-    let gs ← unfoldCases g₂.mvarId g₂.fields[2]!.fvarId!
+    g₁.mvarId.setUserName (.str userNamePre s!"{counter}")
+    let gs ← unfoldCases g₂.mvarId g₂.fields[2]!.fvarId! userNamePre (counter+1)
     return g₁.mvarId :: gs
   catch _ => return []
 
 /-- Implementation of the `fin_cases` tactic. -/
 partial def finCasesAt (g : MVarId) (hyp : FVarId) : MetaM (List MVarId) := g.withContext do
   let type ← hyp.getType >>= instantiateMVars
+  let userName : Name := match (← getMCtx).findDecl? g with
+    | some decl => decl.userName
+    | none => .anonymous
```

Comment:
Last thing, these lines should be replaceable by
```suggestion
  let userName : Name ← g.getTag
```

Thanks! I'll delegate merging to you.
bors d+